### ISSUE TITLE
Markers And PlayerMarkers not anchored properly in topdown view.

### ISSUE THREAD

**Issue by gaupa95** [mapcrafter/mapcrafter#198]
I am currently running a small dedicated server with mapcratfter for my main world.
The problem is that when I view my topdown overview of the map markers are shown to be around 1000 blocks off in x axis.

[Isometric view with working markers](http://i.imgur.com/0O0aFf1.jpg) 

[TopDown View with markers not properly anchored properly](http://i.imgur.com/n7apnw4.jpg)
(Red Arrow indicates where markers should be)

[Mapcrafter Config File](http://pastebin.com/08NyL3ci)

As I've been tinkering a lot I cant say quite what change could have caused this because it was working before.
But I did add render crop, changed image_format And have changed world name since last time I can confirm it working properly.

Do any of you have any idea what could be causing this issiue?

**Comment by m0r13** [comment]
I see you are using a cropped world. Cropped world sometimes get centered using tile offsets, maybe those tile offsets are not correctly considered when calculating the marker positions. Can you try if the markers are correct if you also render your world without cropping?

**Comment by gaupa95** [comment]
Ok. I did just change the settings for the render to not crop the map. 
(removing crop_center_x, crop_center_z and crop_radius from the [global:world] settings) 


Then I re-rendered both the topdown and Isometric view using -r --render-force.
After I ran mapcrafter_markers using the same config file as mapcrafter.

On the Isometric View markers are now off. ca 500 blocks in x axis, but the marker cordinates are corrct. thus the image tiles must be off.

The topdown view are now worse than before. with both map and markers being off in x axis. 

I continued by deleting both maps entierly (including index.html and js. files) As I remembered reading in the documentation that changing the center of the map requires this.

Then rerunning mapcrafter with the same config. 

After waiting the 50min it takes to re-render the Whole world I came back to a blank web site (the index.html).
Apperently I now lack the output_folder/static/ folder. which means index.html cant load all the js files it needs. I need to come back to this tomorrow when I wake up.

**Comment by m0r13** [comment]
Alright then. I'll try to reproduce the problem at home and solve the bug. The tile offsets should work independent of the projection type, but apparently that is not. ;)

I'm sorry for your trouble, easier would have been to create a new world section without the cropping and then using that in another map section.

**Comment by gaupa95** [comment]
No worries. 
The server is rarely played on anyways so no harm done.
I personally just have this to setup and configure a server setup a bit out of the ordinary.
And the fact is I really enjoy using mapcrafter compared to other map generating scripts/programs I've used in the past. 
The amount of custom settings and possibilities are amazing. Keep up the good work man!

well now I need to get back on working on setting up separate high quality views of my towns. 

On another note, have you considered a daemon version of the program? I've been trying to make the program keep the same PID constantly, but I have troubles forking it. 
It's not a problem by any means, I'm just bothered by the 40 000 different logs it creates when its constantly restarting to check/update the map.  just trowing it out there.

**Comment by m0r13** [comment]
Oh I see that you are using a different tile width! This is not properly regarded, I'm trying to fix it right now.

Thanks that you like Mapcrafter! I'm always happy to see people using it and that I'm being able to help users. :)

About the daemon thing: What exactly do you want Mapcrafter to be like a daemon? Do you just want it to run repeatedly? That's exactly what you easily do with a shell script / cron job. If you don't need the info log information, you can disable that in the config file or pipe the (standard) output to /dev/null, you will still get errors and warnings with that as they are printed to the standard error output. Feel free to join me in IRC in freenode (#mapcrafter) if you have questions.

**Comment by gaupa95** [comment]
ah yes, I already have Mapcrafter running repeatedly using systemd (ArchLinux), And the log files are written to systemd's own log. And that flushes at a certain storage limit. As I said it's not a problem. All I really wish to accomplish is that the service uses the same PID each time it restarts.

**Comment by m0r13** [comment]
For the logging ([logging documentation](http://docs.mapcrafter.org/builds/stable/logging.html)): Create a ``log:output`` section in your config file and set the verbosity to ``DEBUG`` or ``WARNING`` if the log output is really a problem for you.

Why do you need Mapcrafter to have the same PID every time? Doesn't systemd find the pid itself when it's starting a program? I don't think it is easy / wanted to change the PID of a program.

**Comment by m0r13** [comment]
I guess you have Mapcrafter built from source or build with the AUR package? Can you pull the source / rebuild or rebuild the AUR package and try if works now?

**Comment by gaupa95** [comment]
ok, pulled the newest source from github, rebuild and installed it.
Then proceeded to remove the TopDown world from my web server.
After re-rendering the Topdown view has the tile info right, but the markers are still off in the X axis, and it seems to be off by a bigger amount now.

Mapcrafter are now returning "Segmentation fault ( core dumped ) after generating the map.

Since I saw the change was in the web UI and web configure ( I think at least?) I then proceeded to delete index.html, settings.js and the static folder. I then moved the static folder from mapcrafter-master.zip (src/data/static) into my template folder and let mapcrafter run again.
I also renamed the old index.html and added the new one from mapcrafter-master.zip into the custom template folder.

Now the map viewer lacks UI features (view select, zoom, rotation view)?
Firefox shows them loaded but leaflet.js responds with 
"Error: Invalid LatLng object: (NaN, NaN)"

Tomorrow I will get back to it, I will try removing mapcrafter from the server, then re-install it. Also going to remove the template folder from the server and delete the whole output folder. but now I need to get to bed. gotta go to work in 3 hours -.-.

**Comment by m0r13** [comment]
You don't have to touch the index.html, the settings.js and static/ in the output directly at all. Mapcrafter is handling all those files. I'm not sure if you did that but just copying the index.html and settings.js from the template directory will break your "UI features" because Mapcrafter adds information regarding to your map to these files.

Just install Mapcrafter as you installed it before (AUR package or from source?) and run it again for your map. Usually you don't need to update any map image files now as I changed only web ui stuff, you just need to update the template files by skipping all maps with the ``-r`` option. But I'm not sure if your settings.js and everything in the output directory is still valid, so better just run Mapcrafter with the config file and without any other settings (except thread count etc. maybe).

**Comment by gaupa95** [comment]
ok, I've now reinstalled mapcrafter (this time from AUR), then removed the custom template folder from the config.
After I deleted index.html and settings.js in the output folder. 
Then I ran mapcrafter again.
Mapcrafter is not returning the segmentation error anymore, And the web ui is now working.
Isometric view is now normal. But the topdown view is really odd. 

I tried to delete the topdown world folder and reran mapcrafter (with only the config file and -j option)
but it did not seem to help.

markers are waaaay more off than before now. 
http://i.imgur.com/YmQJ1fK.jpg
but the issiue dosn't seem to be the markers now. 
It seems like the tile info is wrong.
say I have this isle I use for a referance. it sits on x-2200 z480 y64, both in game and on the isometric view.
but on the topdown view its now at x-170 z2550 y64.

I have no idea why this is.

**Comment by gaupa95** [comment]
Well. For now I'm just removing tile-width for the world. it works now.

**Comment by m0r13** [comment]
I am sorry that the marker positions aren't correct yet. This is because apparently the Minecraft coordinates <--> Leaflet coordinates conversion is not correct. From what I reproduced at home I was able to fix the issue, but maybe there is something I am overlooking. I'll have a look at it again.

**Comment by m0r13** [comment]
When you update the map in this case, just run ``mapcrafter -c myconfig.conf -r``. The ``-r`` makes Mapcrafter just update all the template files because for this way of fixing the issue there are no tile changes needed. Also you really never have to touch the index.html and settings.js in the output folder.

Is your rendered Mapcrafter online somewhere where I could have a look at it myself?

**Comment by gaupa95** [comment]
ok. I can do that. Since I already have the world rendered without tile width (and that works) I'm just renaming the folder containing image data for the TopDown view.

Then I added tile_width=3 to the config. 
[Config](http://pastebin.com/mNtNsqUD) 

After re-rendering the TopDown view again with tile_width enabled (it only takes ca 10min with tile_width enabled) 

If you wish to see for yourself the map is hosted on my local server [here](http://ggwp.project.li/FastWorld/index.html) With the isometic view called "3D" and the topDown View called "Overview"
As stated before the isometric view works as expected.

**Comment by m0r13** [comment]
I just had a look at your map and the header of the index.html tells me you are using Mapcrafter revision 1e5df8c? That's an older version which does not have fixes included yet. Are you sure that you updated Mapcrafter accordingly and used the new version?

**Comment by nheir** [comment]
There is some inconsistency with IsometricRenderView/TopdownRenderView mcToLatLng/latLngToMC.

This patch seems to solve the issue

`diff --git a/src/data/template/static/js/mapcrafterui.js b/src/data/template/static/js/mapcrafterui.js
index 7a1f815..f1aa97f 100644
--- a/src/data/template/static/js/mapcrafterui.js
+++ b/src/data/template/static/js/mapcrafterui.js
@@ -131,13 +131,13 @@ var TopdownRenderView = {
                var lng = 0.5 + (x - tileOffset[0]*tileWidth*16) * block;
                var lat = 0.5 + (z - tileOffset[1]*tileWidth*16) * block;
                
-               var size = mapConfig.tileSize / tileWidth * Math.pow(2, lmap.getZoom());
+               var size = mapConfig.tileSize * Math.pow(2, lmap.getZoom());
                return lmap.unproject([lng * size, lat * size]);
        },
        
        latLngToMC: function(latLng, y, lmap, mapConfig, tileOffset, tileWidth) {
                var point = lmap.project(latLng);
-               var size = mapConfig.tileSize / tileWidth * Math.pow(2, lmap.getZoom());
+               var size = mapConfig.tileSize * Math.pow(2, lmap.getZoom());
                
                var lat = point.y / size;
                var lng = point.x / size;
`

**Comment by m0r13** [comment]
That doesn't work too? If you do a git blame on that file, you can maybe see that I tried some things too but wasn't successfull :/

**Comment by nheir** [comment]
In fact, their is two issues.
 * markers are not at good position
 * tiles are not at good position

For exemple, I set a marker at -180 500 on a map, the mouse UI said the -180 500, tile supposed to be at this place is not.

I manually removed tileoffset and tilewidth from mapcrafterui.js and the result is coherent.

**Comment by m0r13** [comment]
Hmm I see. I have to admit that I kinda hacked the topdown view projection together not completely understanding everything, so there might be an issue like you mentioned it in it. Do you maybe want to have a closer look at it? Unfortunately I'm too busy with a lot of other things recently, so I can't promise when I will find the time to have a closer look at it.

**Comment by m0r13** [comment]
I just did a short test. Am I correct that the problem only occurs with the topdown render view?

**Comment by m0r13** [comment]
@nheir can you test if it works for you now? I re-implemented the coordinate conversion for the topdown view and with my local test-worlds it fixed the problem.

**Comment by m0r13** [comment]
It's not a nightly build yet. I'm still working on the isometric view (as I saw that there are problems too), but I'm getting first improvements there too! I will build some nightlies later today when I'm finished working on the isometric view, so you can test it too.

Don't worry with the hosting. I have a local test-map too.

**Comment by sylcai** [comment]
Hurray! Your latest update DID indeed FIX the issue with the images on a FLAT render being off co-ordinates.

HOWEVER, my cave rendered map still has issues with co-ordinates. I am using an old website to upload the map now, (for my friends that don't live in my house too), and I'll post it here so you can see that the shot of the cave view in my earlier comment is still unchanged and 'broken'.

**Comment by m0r13** [comment]
My latest update? Do you mean the nightlies or the commit I pushed to master? Because the nightly shouldn't have fixed anything.

**Comment by sylcai** [comment]
The commit from 40 minutes ago in mapcrafterui.js

https://github.com/mapcrafter/mapcrafter/blob/master/src/data/template/static/js/mapcrafterui.js

**Comment by m0r13** [comment]
Okay I got to refactor the isometric render view coordinate conversion too. It looks much better now too, though markers always seem to be x-translated by `(tile width - 1) / tile width` tiles. I'm doing further investigations.

**Comment by nheir** [comment]
Ok, i did a bunch of tests with many tile_width, in topdown and iso view.
See the result here : http://died.re/test/ with http://died.re/test/test.conf as configuration

First, mapcrafter doesn't like big tile_width and some tiles render incorrectly so the map is cropped : http://died.re/test/#iso_reallybigtile2/0/0/-66/274/64 

Secondly, tileOffsets is weird, i can't grasp the way it works.
I substituted `(tile width - 1) / tile width * mapConfig.tileSize` by ` -mapConfig.textureSize*16`.

Looking at iso, the offset is right. Then
 * for iso_small and iso_medium, resp. tile_width 2 and 3, an additional shift is required : `(1,-2) * mapConfig.textureSize*32*(tileWidth-1)`
 * for iso_big…reallybig3, resp. tile_width 4, 4, 5 and 7, the additional shift is `(1,-1) * mapConfig.textureSize*32*(tileWidth-1)`
 * for iso_reallybigtile2 of tile_width 8, the additional shift is `(1,0) * mapConfig.textureSize*32*(tileWidth-1)`
Some kind of logarithm of tile_width looks useful but no reason to actually believe in it.

I don't understand how the tile_offsets is defined though.

**Comment by m0r13** [comment]
Damn, I see there are quite some issues with those options. Guess I was too fast with the new version then...

I'll try to have a look at those cropped maps tomorrow. For the topdown view the chunk/tile mapping shouldn't be too difficult to debug. Maybe that's even the source of the other problems with the markers.

**Comment by m0r13** [comment]
Btw the tile offsets are used for cropped worlds. All the tiles are just translated by the middle point so that there is not just a little part of the world very far away from the center (= which would require a high zoom level), but instead that little part is in the middle of the map.

**Comment by m0r13** [comment]
I found a possible problem (rounding wasn't towards -infinity) in the topdown mapping which fixed the problem in my test-world where tiles are missing. Does that work for your topdown maps too?

**Comment by m0r13** [comment]
And I got another one... The cropping might be fixed now. It would be cool if you can re-render your test-map @nheir. Best thing would be to remove the output directly completely as the tile mapping has slightly changed in the isometric render view to prevent weird browser cache problems.

**Comment by nheir** [comment]
Using the last commit, i produce this maps http://died.re/test2/
I can't re-render the previous map for now so i rendered another one a configuration close enough.
Tiles are not cropped and the offsets is almost perfect :D

**Comment by m0r13** [comment]
That's great! What do you mean with almost perfect? I realize that markers should point at the center of blocks, but otherwise the markers are working now, right?

**Comment by nheir** [comment]
I get https://ptpb.pw/1mik.png

**Comment by nheir** [comment]
I confirm the same offset with the previous map and the fix.

**Comment by m0r13** [comment]
I see that there was a single tile offset in horizontal direction in the old implementation too, apparently one tile, but not ``mapConfig.textureSize * 16`` which is only a half tile in isometric render view. That's why I'm a bit wondering, but okay maybe that magic solves it.

With your last message you mean that the fix worked for the other map too?

**Comment by nheir** [comment]
Yes !

**Comment by m0r13** [comment]
Very good then. I applied the fix and confirm that it works for me too.

Without fix:
![screen](https://cloud.githubusercontent.com/assets/1788152/16601838/c96f4a6a-430f-11e6-93f6-e5243ccd6dd4.png)

With fix:
![screen](https://cloud.githubusercontent.com/assets/1788152/16601865/ef455298-430f-11e6-80a4-7584a3743acb.png)

I also improved the positioning of markers a bit by letting them point to the center.

I guess we can call this issue closed then! :+1:
